backend/entityservice/database/util.py

```diff
@@ -145,3 +147,41 @@ def execute_returning_id(cur, query, args):
     else:
         resource_id = query_result[0]
         return resource_id
+
+
+def binary_format(stream: BytesIO):
+    """
+    parse binary format of Postgresql. This is a generator which yields one row at a time.
+    :param stream: a byte stream containing the result of a binary sql query.
+    :return: yield row of the result
+    """
+    stream.seek(0)
+    # Need to read/remove the Postgres Binary Header, Trailer, and the per tuple info
+    # https://www.postgresql.org/docs/current/sql-copy.html
+    header = stream.read(15)
+    assert header == b'PGCOPY\n\xff\r\n\x00\x00\x00\x00\x00', "Invalid Binary Format"
+    header_extension = stream.read(4)
+    assert header_extension == b'\x00\x00\x00\x00', "Need to implement skipping postgres binary header extension"
+    # now iterate over the rows
+    while True:
+        num_elements = int.from_bytes(stream.read(2), byteorder='big')  # apparently bytes are in network byte order
+        if num_elements == 65535:   # we reached the end of the results. Binary trailer of \xff\xff
+            return
+        row_values = []
+        for i in range(num_elements):
+            size = int.from_bytes(stream.read(4), byteorder='big')
+            el_value = stream.read(size)
+            row_values.append(el_value)
+        if num_elements == 1:
+            yield row_values[0]
+        else:
+            yield row_values
+
+
+def compute_encoding_ids(entity_ids: Iterable[int], dp_id: int):
+    """ compute unique encoding ids for given entity ids
+    The user provides entity ids. Although unique for the user, different user can have the same entity ids.
+    However, we want to use the id as a unique identifier for the encoding in the database. Thus, we create a
+    unique number here. """
+    dp_offset = dp_id << 32
+    return [dp_offset + entity_id for entity_id in entity_ids]
```

Comment:
Does the `entity_id` need to be encoded into the `encoding_id`? If not perhaps it would be better to simply insert the encodings and have the database assign unique row keys?

Reply:
If you let the db assign unique ids, we run into a problem: First some task inserts the encodings. This is fine. At some later stage, another task wants to grab some of the encodings. This tasks does not know the unique ids, thus has to query them with 'encoding id' and 'dp id'. To speed this up, we would have to create an index for (encoding_id, dp_id) on the encodingblocks table in the database. In my tests, this index creation was the main culprit for the mediocre performance. Once created, everything's nice and fast, but during creation everything grinds to a halt.  
I couldn't figure out how to speed up the index creation. That's why I introduced this unique ids to get around the index.

Reply:
> This tasks does not know the unique ids, thus has to query them with 'encoding id' and 'dp id'. To speed this up, we would have to create an index for (encoding_id, dp_id) on the encodingblocks table in the database.

Ahh right I'm with you, I'd forgotten we now might be selecting encodings by multiple blocks at once or as a chunk within one block.

> In my tests, this index creation was the main culprit for the mediocre performance. Once created, everything's nice and fast, but during creation everything grinds to a halt.
I couldn't figure out how to speed up the index creation.

That's a shame, I'd have thought it could be done but might not be worth the added complexity. E.g. inserting into a temporary table without an index, adding the index and then copying into the real table. 

I guess the main alternative is a table mapping encoding_ids to a work package?

Anyhow the unique ids make more sense now